File: lib/website.js
var db = require('../lib/database');
var logger = require('../lib/logger');
var httpcodes = require('../lib/http-status-codes');
var pushBullet = require('pushbullet');
var request = require('request');

function Website(id, protocol, url) {
	this.id = id;
	this.protocol = protocol;
	this.url = url;

	Website.prototype.runCheck = function(allowSecondTry) {
		request({
			url: protocol + '://' + url,
			headers: {'User-Agent': 'UpAndRunning (https://github.com/MarvinMenzerath/UpAndRunning)'}
		}, function(error, response) {
			var status;
			if (response === undefined) {
				status = 'Server not found';
			} else {
				status = response.statusCode + ' - ' + httpcodes[response.statusCode];
			}

			// get website's name and previous status
			var name;
			var oldStatus;
			var allowPush = true;
			if (global.PBAPI != "") {
				db.query("SELECT name, status FROM website WHERE id = ?", [id], function(err, rows) {
					if (err) {
						logger.error("Unable get website's name and old status: " + err.code);
						allowPush = false;
					} else {
						name = rows[0].name;
						oldStatus = rows[0].status;
					}
				});
			}

			if (status.indexOf(200) > -1 || status.indexOf(301) > -1 || status.indexOf(302) > -1) {
				// success
				db.query("UPDATE website SET status = ?, time = NOW(), ups = ups + 1, totalChecks = totalChecks + 1 WHERE id = ?;", [status, id], function(err) {
					if (err) {
						logger.error("Unable to save new website-status: " + err.code);
					} else {
						calcAvgAvailability();
					}
				});
			} else {
				// failure
				if (allowSecondTry) {
					setTimeout(function() {
						new Website(id, protocol, url).runCheck(false);
					}, 1000);
					allowPush = false;
				} else {
					db.query("UPDATE website SET status = ?, time = NOW(), lastFailStatus = ?, lastFailTime = NOW(), downs = downs + 1, totalChecks = totalChecks + 1 WHERE id = ?;", [status, status, id], function(err) {
						if (err) {
							logger.error("Unable to save new website-status: " + err.code);
						} else {
							calcAvgAvailability();
						}
					});
				}
			}

			// compare previous and new status and send a push if there is a change
			if (global.PBAPI != "" && allowPush && name != undefined && oldStatus != undefined) {
				setTimeout(function() {
					if (oldStatus != status) {
						var pusher = new pushBullet(global.PBAPI);
						pusher.note({}, global.TITLE + " - Status Change", name + " (" + url + ") went from \"" + oldStatus + "\" to \"" + status + "\".", function(err) {
							if (err) {
								logger.error("Unable to send PushBullet-notification: " + err);
							}
						});
					}
				}, 2000); // wait until you have all the data (dirty fix)
			}
		});

		function calcAvgAvailability() {
			db.query("SELECT ((SELECT ups FROM website WHERE id = ?) / (SELECT totalChecks FROM website WHERE id = ?))*100 AS avg", [id, id], function(err, rows) {
				if (err) {
					logger.error("Unable to calculate new website-availability: " + err.code);
					return;
				}
				var avgAvail = rows[0].avg.toFixed(2);

				db.query("UPDATE website SET avgAvail = ? WHERE id = ?;", [avgAvail, id], function(err) {
					if (err) {
						logger.error("Unable to save new website-availability: " + err.code);
					}
				});
			});
		}
	};
}

module.exports = Website;

File: lib/website.go
package lib

import (
	"database/sql"
	"github.com/franela/goreq"
	"github.com/op/go-logging"
	"strconv"
	"time"
)

// Represents a single Website-object.
type Website struct {
	Id          int
	Protocol    string
	Url         string
	CheckMethod string
}

// Runs a check and saves the result inside the database.
func (w *Website) RunCheck(secondTry bool) {
	// Request new Status
	var requestStartTime = time.Now()
	res, err := goreq.Request{Uri: w.Protocol + "://" + w.Url, Method: w.CheckMethod, UserAgent: "UpAndRunning2/" + GetConfiguration().Static.Version + " (https://github.com/MarvinMenzerath/UpAndRunning2)", MaxRedirects: GetConfiguration().Dynamic.Redirects, Timeout: 5 * time.Second}.Do()
	var requestDuration = time.Now().Sub(requestStartTime).Nanoseconds() / 1000000

	var newStatusText string
	var newStatusCode int
	if err != nil {
		if secondTry {
			newStatusText = "Host not found"
			newStatusCode = 0

			// On Timeout: allow second try
			if serr, ok := err.(*goreq.Error); ok {
				if serr.Timeout() {
					newStatusText = "Timeout"
					newStatusCode = 0
				}
			}
		} else {
			time.Sleep(time.Millisecond * 1000)
			w.RunCheck(true)
			return
		}
	} else {
		newStatusText = GetHttpStatus(res.StatusCode)
		newStatusCode = res.StatusCode
		defer res.Body.Close()
	}

	w.sendNotifications(newStatusCode, newStatusText)

	// Save the new Result
	_, err = db.Exec("INSERT INTO checks (websiteId, statusCode, statusText, responseTime, time) VALUES (?, ?, ?, ?, NOW());", w.Id, newStatusCode, newStatusText, requestDuration)
	if err != nil {
		logging.MustGetLogger("").Error("Unable to save the new Website-status: ", err)
		return
	}
}

// Gets the notification-settings and sends a notification (if necessary and requested)
func (w *Website) sendNotifications(newStatusCode int, newStatusText string) {
	var (
		pushbulletKey string
		email         string
		name          string
		oldStatusCode string
		oldStatusText string
	)

	db := GetDatabase()
	err := db.QueryRow("SELECT pushbulletKey, email FROM notifications WHERE websiteId = ?", w.Id).Scan(&pushbulletKey, &email)
	if err != nil {
		if err == sql.ErrNoRows {
			return
		}
		logging.MustGetLogger("").Error("Unable to get Website's notification-settings: ", err)
	}

	// Check for empty result
	if pushbulletKey == "" && email == "" {
		return
	}

	err = db.QueryRow("SELECT name FROM websites WHERE id = ?", w.Id).Scan(&name)
	if err != nil {
		logging.MustGetLogger("").Error("Unable to get Website's data: ", err)
		return
	}
	err = db.QueryRow("SELECT statusCode, statusText FROM checks WHERE websiteId = ? ORDER BY id DESC LIMIT 1", w.Id).Scan(&oldStatusCode, &oldStatusText)
	switch {
	case err == sql.ErrNoRows:
		return
	case err != nil:
		logging.MustGetLogger("").Error("Unable to get Website's data: ", err)
		return
	}

	oldStatus := oldStatusCode + " - " + oldStatusText
	newStatus := strconv.Itoa(newStatusCode) + " - " + newStatusText
	if oldStatus != newStatus {
		if pushbulletKey != "" {
			sendPush(pushbulletKey, name, w.Url, newStatus, oldStatus)
		}
		if email != "" {
			sendMail(email, name, w.Url, newStatus, oldStatus)
		}
	}
}
